### ISSUE TITLE
Reddit Builds a House

### ISSUE THREAD

**Issue by demipixel** [mhsjlw/flying-squid#31]
The original: https://www.reddit.com/r/Minecraft/comments/3ilm3b/reddit_builds_a_house/

The idea: https://www.reddit.com/r/Minecraft/comments/3iz6fw/help_plan_reddit_builds_a_house_2/

**The basic idea:**

- Empty terrain on server (at least something to build off of), ~20 slots on server
- Player joins and has three minutes to place a block. They can edit terrain but it will only be displayed locally.
- Use /reset (or something) to reset all blocks (removes all local blocks).
- Use /confirm to get a confirmation message for the latest block placed (e.g. "Place block X at 5, 63, 4? Use /confirm yes/no")
- Using /confirm yes will "place" the block and all users will see it (and it will be saved to the world). The user will be kicked with the message "Join back to spectate!"
- If you join and have already placed a block, you will be placed in a spectator mode. If you have been on for over 60 seconds and someone is trying to join, you will be kicked

The exact "slot management" system is rough since we only have 20 slots and I'm sure a LOT more people will be coming than that, however this should give a basic idea of what I'm trying to do.

**TL;DR 3 minutes to place block, local editing until /reset or /confirm, join back to spectate temporarily**

*I'm assuming the local place/destroy will be the hardest part of this*

**Comment by rom1504** [comment]
Okay, so with the current architecture + #29 , you'd do this : 
* config file : disable placeBlock and digging plugins so people can edit the terrain
* external plugin : /reset resend the "empty" world to the client
* external plugin : store place block in a fifo structure
* external plugin : /confirm does what you said
* external plugin : implement the spectator on join

I don't see any need for cancelling anything here. Maybe it would be interesting to emit a blockPlace event a little bit smarter than the raw block_place packet, but that's all.

**Comment by mhsjlw** [comment]
@rom1504 Implementing spectator is as easy as storing them in a JSON file then using the game_state_change with reason of 3 then gamemode 3

**Comment by rom1504** [comment]
That's not how you'd do it here though, since it needs to be dynamic ;)

**Comment by demipixel** [comment]
@rom1504 Yeah, we'd basically need to completely overwrite the blockPlace because we never want to save it to the world unless they do /confirm

And /reset doesn't send an empty world but the original world (just to make sure we're clear)

Also, does it need to be 4 external plugins? That seems like it would be difficult to organize, why couldn't it be one?

**Comment by rom1504** [comment]
No I meant 1 external plugin yeah, not 4, that was just 4 things the external plugin need to do.

**Comment by demipixel** [comment]
Would it be good to make some sudo code in order to better demonstrate plugins? (Others could make them as well so we can compare how plugins would look)

**Comment by rom1504** [comment]
what do you mean by "sudo code" ?

**Comment by mhsjlw** [comment]
He means pseudo code, it is example code that expresses your ideas but doesn't actually work when you attempt to compile it, @rom1504 

Google definition, "Pseudocode is an informal high-level description of the operating principle of a computer program or other algorithm. It uses the structural conventions of a programming language, but is intended for human reading rather than machine reading."

**Comment by rom1504** [comment]
ah okay. Yeah well why not demipixel, but I'm not sure if writing it in pseudocode is easier than in js, but feel free to.

**Comment by demipixel** [comment]
Well I'll just write in js that wouldn't run

**Comment by demipixel** [comment]
How it would work:

https://gist.github.com/demipixel/301537f68b0aeec5eb83

This is more for "how plugins would work in general".

**Comment by rom1504** [comment]
@demipixel put your plugin in a repo and we can close this ;)

**Comment by rom1504** [comment]
http://mammal.io/articles/using-es6-today/ is a simple way to add es6 (without having to put that whole gulp thing). Might be useful for plugins, I'll try to set it up for flying-squid-irc to show it's not complicated.

**Comment by demipixel** [comment]
This should be closed. The project isn't complete, but the issue related to completing the project is.
